### PR TITLE
[SPARK-37425][PYTHON] Inline type hints for python/pyspark/mllib/recommendation.py

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Inline type hints for recommendation in python/pyspark/mllib/

### Why are the changes needed?
We can take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by dchvn** [comment]
Gentle ping @zero323

**Review by itholic — APPROVED** [review]
Looks pretty good

**Review by zero323 — COMMENTED** [review]
This 

https://github.com/apache/spark/blob/34ac04e9fb4b178ba6d34a1283ffa3b3f75fd0f0/python/pyspark/mllib/recommendation.py#L32


should be expressed with typed `NamedTuple`.

File: python/pyspark/mllib/recommendation.py [hunk 1]
```diff
@@ -245,8 +249,15 @@ def _prepare(cls, ratings):
...
-    ):
+        cls,
+        ratings: Union[RDD[Rating], RDD[Tuple[int, int, float]]],
+        rank: int,
+        iterations: int = 5,
+        lambda_: float = 0.01,
+        blocks: int = -1,
+        nonnegative: bool = False,
+        seed: Optional[int] = None,
+    ) -> "MatrixFactorizationModel":
```
Comment (zero323) [inline python/pyspark/mllib/recommendation.py:260]: There should be no need for quotes here.

File: python/pyspark/mllib/recommendation.py [hunk 2]
```diff
@@ -296,15 +307,15 @@ def train(
...
-    ):
+        ratings: Union[RDD[Rating], RDD[Tuple[int, int, float]]],
+        rank: int,
+        iterations: int = 5,
+        lambda_: float = 0.01,
+        blocks: int = 1,
+        alpha: float = 0.01,
+        nonnegative: bool = False,
+        seed: Optional[int] = None,
+    ) -> "MatrixFactorizationModel":
```
Comment (zero323) [inline python/pyspark/mllib/recommendation.py:318]: Ditto./

File: python/pyspark/mllib/recommendation.py [hunk 3]
```diff
@@ -296,15 +307,15 @@ def train(
...
-        blocks=-1,
-        alpha=0.01,
-        nonnegative=False,
-        seed=None,
-    ):
+        ratings: Union[RDD[Rating], RDD[Tuple[int, int, float]]],
+        rank: int,
+        iterations: int = 5,
+        lambda_: float = 0.01,
+        blocks: int = 1,
```
Comment (zero323) [inline python/pyspark/mllib/recommendation.py:314]: This should be `blocks: int = -`1,`

File: python/pyspark/mllib/recommendation.py [hunk 4]
```diff
@@ -224,7 +228,7 @@ class ALS:
     """
 
     @classmethod
-    def _prepare(cls, ratings):
+    def _prepare(cls, ratings: Any) -> RDD:
```
Comment (zero323) [inline python/pyspark/mllib/recommendation.py:231]: Returned type should be properly annotated with `RDD[Rating]`.

We might also annotate `ratings` as `Union[RDD[???], DataFrame])`, but it is OK to have `Any` in this case.

**Comment by dchvn** [comment]
Updated, thanks @zero323

**Comment by itholic** [comment]
@zero323 thanks for catching the detail! LGTM

**Comment by zero323** [comment]
Merged into master.

**Pull request closed without merging**